share/new_eds/close_once.go

```diff
@@ -0,0 +1,78 @@
+package eds
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/shwap"
+)
+
+var _ AccessorCloser = (*closeOnce)(nil)
+
+var errAccessorClosed = errors.New("accessor is closed")
+
+type closeOnce struct {
+	f      AccessorCloser
+	closed atomic.Bool
+}
+
+func WithClosedOnce(f AccessorCloser) AccessorCloser {
+	return &closeOnce{f: f}
+}
+
+func (c *closeOnce) Close() error {
+	if !c.closed.Swap(true) {
+		err := c.f.Close()
+		// release reference to the accessor to allow GC to collect all resources associated with it
+		c.f = nil
+		return err
+	}
+	return nil
+}
+
+func (c *closeOnce) Size(ctx context.Context) int {
```

Comment:
Random realization that we could fix in some other PR: if it takes context, it should return an error